tahowallet/dapp
# Generate and claim XP

Resolves https://github.com/tahowallet/dapp/issues/179

### What

This PR let's you simulate XP drops and test claiming XP that is available for a given user.

#### Technical details:
- to test this PR you need fork contracts from Arb Sepolia
- `deployment-info.json` here matches Arb Sepolia contracts
- we have 2 new scripts to make testing possible:
  - `yarn generate-xp` will generate file with a merkle tree of XP - the data is fake and you can adjust the mocked data [here](https://github.com/tahowallet/dapp/pull/303/files#diff-617dcabec8a4218045c2146390cdd56e2fdf1d57d96813b5c2fae944b60369c6R5-R18)
  - once you have generated xp you can make the XP drop on your local hardhat fork by adjusting variables [here](https://github.com/tahowallet/dapp/pull/303/files#diff-1b8d7c3c11e98f45d654e1c016e6668fc7c340f8304d025d0d6cbd516fcf00abR4-R10) and running `yarn distribute-xp`


#### XP drops merkle data structure

Dapp assumes that all xp drops merkle trees will be stored in the:
`data/xp/<realmID>/xp_<realmID>_<dropIndex>.json`

Dapp assumes there will be a leaderboard merkle tree file that will accumulate all XP drops for a given realm in one tree and store this file in the:
`data/xp/<realmID>/leaderboard.json`
Leaderboard data is not posted on-chain, it is used internally by the dapp. This PR doesn't include the script to generate accumulated XP leaderboard data, existing files are just mocks.

### Testing steps

XP
1. Use one of the merkle trees from `data/xp` or override them by your own, generated by the script
2. Update vars in `distribute-xp` script (you have to figure out the realm address for a given realm id - easy redux store)
3. Run hardhat localhost Arb Sepolia fork
4. Run `distribute-xp` script against the fork (you can distribute one or more drops - on one realm or multiple)
5. Connect in the dapp to one of the accounts that were in the drop
6. Account has to be staked somewhere to see the banner with claimable XP rewards
7. See if the amount to claim is correct
8. Try to claim (TODO - will fail, need to check what is wrong with merkle tree because it is failing on the merkle proof)


Leaderboard
1. See if leaderboard is connecting to the correct data from `leaderboard.json` files
2. See if leaderboard is not failing if file is missing
